### ISSUE TITLE
The module doesn't work if ipset in use

### ISSUE THREAD

**Issue by anubisdonetks** [voxpupuli/puppet-ipset#68]
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 5.5.14
- Ruby: ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux-gnu]
- Distribution: Ubuntu 22.04 LTS
- Module version: 1.2.3 (I see this bug in code 2.1.0) 

## How to reproduce (e.g Puppet code you use)
Create ipset and add iptables rule for use this ipset
Example:
`iptables -A INPUT -m set --match-set test src -j ACCEPT`

## What are you seeing
We found that when using the ipset firewall or ubuntu 20 and older, the address addition does not occur.
The error occurs while running the ipset_sync script

## Output log
ipset v7.15: Set cannot be destroyed: it is in use by a kernel component
ipset v7.15: Error in line 1: Set cannot be created: set with the same name already exists

## Any additional information you'd like to impart
We create new function in ipset_sync
```bash
function import_ipset_new() {
  local id=$1
  local alias=${2:-${id}}
  local no_content=${3:-0}
  local f_content=$(get_content_file ${id})
  if [ ! -f "${f_content}" ]; then
    echo "Set configuration file '${f_content}' does not exist!" >&2
    exit 6
  else
    while read LINE; do
            ipset add ${id} $LINE
    done < ${f_content}
  fi

}
```

and change lines 
`ipset destroy ${set_id}` to `ipset flush ${set_id}`
`import_ipset ${set_id} ${set_id} ${ignore_contents}` to
`import_ipset_new ${set_id} ${set_id} ${ignore_contents}`

**Comment by wavilov** [comment]
Fix works, thank!

**Comment by oliparcol** [comment]
Unless I'm mistaken, the suggested solution allows syncing the content of the *ipset* but doesn't update its settings (e.g. change from `inet` to `inet6`). From what I read, I don't think there is an easy way to update an *ipset* setting if it is currently used by the kernel.

**Comment by weastur** [comment]
Ubuntu 22.04 has a newer version of ipset which has several improvements. The issue we've faced exists because of the following two:

- Expose the initval hash parameter to userspace
- Add bucketsize parameter to all hash types

So now, `ipset_sync` calculate header diff like this:

```shell
# ipset_sync -c '/etc/ipset.d/' -d -v -i wan-networks
1c1
< create wan-networks hash:net family inet maxelem 65536 bucketsize 12 initval 0x112ae629
---
> create wan-networks hash:net family inet maxelem 65536
```

I believe `initval` is useless for this module, but the `bucketsize` might be helpful. Anyway, as far as I don't interested in `bucketsize`, I ignore both with the next changes [here](https://github.com/voxpupuli/puppet-ipset/blob/fc81728c7ac7063da7b1aef7b0fcc11da49fbcf3/files/ipset_sync#L125)

```bash
    <(get_ipset_dump ${id} | grep ^${pfx} | sed 's/hashsize [0-9]\+ //g' | sed 's/\s*bucketsize [0-9]\+\s*//g' | sed 's/\s*initval 0x[a-f0-9]\+\s*//g') \
    <(construct_ipset_dump ${id} | grep ^${pfx} | sed 's/hashsize [0-9]\+ //g' | sed 's/\s*bucketsize [0-9]\+\s*//g' | sed 's/\s*initval 0x[a-f0-9]\+\s*//g') 
```

I could make a PR if this is an acceptable solution, or even add `bucketsize` support depending on `ipset` version.

**Comment by oliparcol** [comment]
There are some open PR that partially solve this issue. We are using the following patch that works on Ubuntu 22.04:

```
diff --git a/files/ipset_sync b/files/ipset_sync
index aefc982..c702577 100755
--- a/files/ipset_sync
+++ b/files/ipset_sync
@@ -122,8 +122,8 @@ function ipset_insync_common() {
   # compare configured and runtime config
   tf=$(mktemp)
   diff \
-    <(get_ipset_dump ${id} | grep ^${pfx} | sed 's/hashsize [0-9]\+ //g') \
-    <(construct_ipset_dump ${id} | grep ^${pfx} | sed 's/hashsize [0-9]\+ //g') \
+    <(get_ipset_dump ${id} | grep ^${pfx} | sed -e 's/ hashsize [0-9]\+//g' -e 's/ initval 0x[0-9a-f]\+//g' -e 's/ bucketsize [0-9]\+//g' -e 's/[[:space:]]*$//') \
+    <(construct_ipset_dump ${id} | grep ^${pfx} | sed -e 's/ hashsize [0-9]\+//g' -e 's/ initval 0x[0-9a-f]\+//g' -e 's/ bucketsize [0-9]\+//g' -e 's/[[:space:]]*$//') \
     > ${tf}
   local rv=$?
 
```

**Comment by maxadamo** [comment]
for clarity, we are talking about PR https://github.com/voxpupuli/puppet-ipset/pull/67

**Comment by maxadamo** [comment]
@oliparcol if we don't get a reply for that PR, I'm gonna create a new one and we can disregard PR 67. 
We need this change in order to use this module with Ubuntu 22.04

**Comment by oliparcol** [comment]
I can also do it, I didn't do it yet as @anubisdonetks might want to update the initial PR.

**Comment by maxadamo** [comment]
sorry for the late reply. This one was fixed with #81 and it can be closed.
